src/zope/hookable/_zope_hookable.c

```diff
@@ -12,153 +12,119 @@
  #
  ############################################################################*/
 
-/* _zope_hookable.c
-
-   Provide an efficient implementation for hookable objects
-
- */
+static char module__doc__[] = (
+    "Provide an efficient implementation for hookable objects"
+);
 
 #include "Python.h"
 #include "structmember.h"
 
-typedef struct {
+typedef struct
+{
     PyObject_HEAD
-    PyObject *old;
-    PyObject *implementation;
+    PyObject* original;
+    PyObject* implementation;
 } hookable;
 
 static int
-hookable_init(hookable *self, PyObject *args, PyObject *kwds)
+hookable_init(hookable* self, PyObject* args, PyObject* kwds)
 {
-  static char *kwlist[] = {"implementation", NULL};
-  PyObject *implementation;
-
-  if (! PyArg_ParseTupleAndKeywords(args, kwds, "O:hookable", kwlist,
-                                    &implementation))
-    return -1;
-
-  Py_INCREF(implementation);
-  Py_INCREF(implementation);
-  Py_XDECREF(self->old);
-  self->old = implementation;
-  Py_XDECREF(self->implementation);
-  self->implementation = implementation;
-
-  return 0;
+    static char* kwlist[] = { "implementation", NULL };
+    PyObject* implementation;
+
+    if (!PyArg_ParseTupleAndKeywords(
+          args, kwds, "O:hookable", kwlist, &implementation))
+        return -1;
+
+    /* Both 'self->original' and 'self->implementation' are originally
+     * set to the passed-in 'implementation', hence the need for
+     * two increfs.
+     */
+    Py_INCREF(implementation);
+    Py_XDECREF(self->original);
+    self->original = implementation;
+
+    Py_INCREF(implementation);
+    Py_XDECREF(self->implementation);
+    self->implementation = implementation;
+
+    return 0;
 }
 
 static int
-hookable_traverse(hookable *self, visitproc visit, void *arg)
+hookable_traverse(hookable* self, visitproc visit, void* arg)
 {
-  if (self->implementation != NULL && visit(self->implementation, arg) < 0)
-    return -1;
-  if (self->old != NULL
-      && self->old != self->implementation
-      && visit(self->old, arg) < 0
-      )
-    return -1;
-
-  return 0;
+    Py_VISIT(Py_TYPE(self));
```

Comment:
@tseaver https://docs.python.org/3.13/howto/isolating-extensions.html#tp-traverse-in-python-3-8-and-lower says this should not happen in Python 3.8 and lower.

Reply:
@davisagli Thanks for the catch!  c3ef500